File: 31_40/SearchInRotatedSortedArray.py
# 33.  Search in Rotated Sorted Array

class Solution:
    def search(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        l = 0
        r = len(nums) - 1
        while l <= r:
            m = int((l + r) / 2)
            if nums[m] == target:
                return m
            # 如果中点值小于右边，则右半边有序
            if nums[m] < nums[r]:
                if target > nums[m] and target <= nums[r]:
                    l = m + 1
                else:
                    r = m - 1
            # 如果中点值大于右边，则左半边有序
            else:
                if target < nums[m] and target >= nums[l]:
                    r = m - 1
                else:
                    l = m + 1
        return -1

File: go_solution/031_040/searchInRotatedSortedArray/searchInRotatedSortedArray.go
package searchInRotatedSortedArray

// 33. Search in Rotated Sorted Array
// 思路：除了比较 nums[mid] 和 target 来决定如何移动 l 和 r，还要判断 nums[mid] 和 target 是不是在同一边
func search(nums []int, target int) int {
	if len(nums) == 0 {
		return -1
	}
	l, r := 0, len(nums)-1
	for l <= r {
		mid := (l + r) / 2
		if target == nums[mid] {
			return mid
		}
		if nums[l] <= nums[mid] {
			if nums[l] <= target && target <= nums[mid] {
				r = mid - 1
			} else {
				l = mid + 1
			}
		} else {
			if target >= nums[mid] && target <= nums[r] {
				l = mid + 1
			} else {
				l = mid - 1
			}
		}
	}
	return -1
}
